### DISCUSSION TITLE
升级Kubevela 版本从1.3.6 到1.5.11 之后 workflow step suspend 一直是running状态，无法继续工作流。

### DISCUSSION THREAD

**Discussion by skyshaokai** [kubevela/kubevela#5908] [Q&A] [answered]
升级Kubevela 版本从1.3.6 到1.5.11 之后 workflow step suspend 一直是running状态，无法继续工作流。
停在 suspend 步骤之后，操作继续没有效果。
steps:
    - firstExecuteTime: "2023-04-24T07:46:46Z"
      id: pxpge97ni7
      lastExecuteTime: "2023-04-24T07:46:46Z"
      name: cpdinf-cloudnative-consumer-jarvis-bd-bddwd
      phase: succeeded
      type: deploy
    - firstExecuteTime: "2023-04-24T07:46:46Z"
      id: lml8tn7y87
      lastExecuteTime: "2023-04-24T07:48:24Z"
      name: suspend-1
      phase: running
      type: suspend
    suspend: true
    terminated: false

      phase 一直是 running 状态。  

操作继续：[{"op":"replace","path":"/status/workflow/suspend","value":false}]
还是一直未暂停状态。

**Comment by wonderflow** [comment]
ping @FogDong 

我估计是这块 Workflow 的resume 逻辑变了，不能通过  `[{"op":"replace","path":"/status/workflow/suspend","value":false}]` 管理了。

**Reply by skyshaokai** [reply]
看了一下1.3.6 的代码如下：
`func resumeWorkflow(kubecli client.Client, app *v1beta1.Application) error {
	// set the workflow suspend to false
	app.Status.Workflow.Suspend = false

	if err := kubecli.Status().Patch(context.TODO(), app, client.Merge); err != nil {
		return err
	}

	fmt.Printf("Successfully resume workflow: %s\n", app.Name)
	return nil
}`

1.5.11 的代码如下：


// ResumeWorkflow resume workflow
func ResumeWorkflow(ctx context.Context, kubecli client.Client, app *v1beta1.Application) error {
	app.Status.Workflow.Suspend = false
	steps := app.Status.Workflow.Steps
	for i, step := range steps {
		if step.Type == wfTypes.WorkflowStepTypeSuspend && step.Phase == common.WorkflowStepPhaseRunning {
			steps[i].Phase = common.WorkflowStepPhaseSucceeded
		}
		for j, sub := range step.SubStepsStatus {
			if sub.Type == wfTypes.WorkflowStepTypeSuspend && sub.Phase == common.WorkflowStepPhaseRunning {
				steps[i].SubStepsStatus[j].Phase = common.WorkflowStepPhaseSucceeded
			}
		}
	}
	if err := kubecli.Status().Patch(ctx, app, client.Merge); err != nil {
		return err
	}
	return nil
}



不清楚这块代码为什么要变化？我先试试按照1.5.11 代码逻辑来操作一下看看是否ok。

**Reply by FogDong** [reply, marked as answer]
> 不清楚这块代码为什么要变化？

在早期版本，suspend 步骤执行后状态为 succeeded，而工作流的状态为暂停。
在 1.5 版本后修改为，在运行 suspend 步骤时，状态为 running（1.8 版本修改为 suspending），同时工作流的状态为暂停。这样能够清楚地看到工作流是被哪个步骤暂停的。

所以不能只用 `[{"op":"replace","path":"/status/workflow/suspend","value":false}]` 来继续工作流，还需要把所有正在 runnning 的 suspend 步骤改为 succeeded。

**Reply by skyshaokai** [reply]
好的。多谢了。
